server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java

```diff
@@ -730,4 +731,104 @@ public void testExtractNestedSources() {
+                }
+
+                nextLayer = value;
+            }
+
+            return value;
+        };
+
+        {
+            XContentBuilder builder = XContentFactory.jsonBuilder().startObject().field("test", "value").endObject();
+
+            Map<String, Object> map = toSourceMap(Strings.toString(builder));
+            assertThat(XContentMapValues.insertValue("test", map, "value2").toString(), equalTo("value"));
+            assertThat(getMapValue.apply(map, "test"), equalTo("value2"));
+            assertThat(XContentMapValues.insertValue("test.me", map, "test.me.value"), nullValue());
+            assertThat(getMapValue.apply(map, "test\\.me"), equalTo("test.me.value"));
+            assertThat(XContentMapValues.insertValue("something.else.2", map, "something.else.2.value"), nullValue());
+            assertThat(getMapValue.apply(map, "something\\.else\\.2"), equalTo("something.else.2.value"));
+        }
+        {
+            XContentBuilder builder = XContentFactory.jsonBuilder().startObject();
+            builder.startObject("path1").startObject("path2").field("test", "value").endObject().endObject();
+            builder.endObject();
+
+            Map<String, Object> map = toSourceMap(Strings.toString(builder));
+            assertThat(XContentMapValues.insertValue("path1.path2.test", map, "value2").toString(), equalTo("value"));
+            assertThat(getMapValue.apply(map, "path1.path2.test"), equalTo("value2"));
+            assertThat(XContentMapValues.insertValue("path1.path2.test_me", map, "test_me_value"), nullValue());
+            assertThat(getMapValue.apply(map, "path1\\.path2\\.test_me"), equalTo("test_me_value"));
+            assertThat(XContentMapValues.insertValue("path1.non_path2.test", map, "test_value"), nullValue());
+            assertThat(getMapValue.apply(map, "path1\\.non_path2\\.test"), equalTo("test_value"));
+
+            assertThat(XContentMapValues.insertValue("path1.path2", map, Map.of("path3", "bar")), equalTo(Map.of("test", "value2")));
+            assertThat(getMapValue.apply(map, "path1.path2"), equalTo(Map.of("path3", "bar")));
+
+            assertThat(XContentMapValues.insertValue("path1", map, "baz"), equalTo(Map.of("path2", Map.of("path3", "bar"))));
+            assertThat(getMapValue.apply(map, "path1"), equalTo("baz"));
+        }
+
+        // lists
+        {
+            XContentBuilder builder = XContentFactory.jsonBuilder().startObject();
+            builder.startObject("path1").array("test", "value1", "value2").endObject();
+            builder.endObject();
+            Map<String, Object> map = toSourceMap(Strings.toString(builder));
+
+            assertThat(
+                XContentMapValues.insertValue("path1.test", map, List.of("value3", "value4", "value5")),
+                equalTo(List.of("value1", "value2"))
+            );
+            assertThat(getMapValue.apply(map, "path1.test"), equalTo(List.of("value3", "value4", "value5")));
+        }
```

Comment:
This test is incomplete, will be adding more cases to it

Reply:
Let's separate each case to its own test function? It's a lot easier to read and reproduce issues when each test case is simple and isolated.